openmetadata-service/src/main/java/org/openmetadata/service/security/auth/BasicAuthenticator.java

```diff
@@ -0,0 +1,477 @@
+      if (!allowedDomains.contains("all") && !allowedDomains.contains(emailDomain)) {
+        LOG.error("Email with this Domain not allowed: " + newRegistrationRequestEmail);
+        throw new BadRequestException("Email with the given domain is not allowed. Contact Administrator");
+      }
+      validateEmailAlreadyExists(newRegistrationRequestEmail);
+      PasswordUtil.validatePassword(newRegistrationRequest.getPassword());
+      LOG.info("Trying to register new user [" + newRegistrationRequestEmail + "]");
+      User newUser = getUserFromRegistrationRequest(newRegistrationRequest);
+      // remove auth mechanism from the user
+      User registeredUser = userRepository.create(null, newUser);
+      registeredUser.setAuthenticationMechanism(null);
+      return registeredUser;
+    } else {
+      throw new CustomExceptionMessage(BAD_REQUEST, SELF_SIGNUP_ERROR);
+    }
+  }
+
+  @Override
+  public void confirmEmailRegistration(UriInfo uriInfo, String emailToken) throws IOException {
+    EmailVerificationToken emailVerificationToken = (EmailVerificationToken) tokenRepository.findByToken(emailToken);
+    if (emailVerificationToken == null) {
+      throw new EntityNotFoundException("Invalid Token. Please issue a new Request");
+    }
+    User registeredUser =
+        userRepository.get(null, emailVerificationToken.getUserId(), userRepository.getFieldsWithUserAuth("*"));
+    if (Boolean.TRUE.equals(registeredUser.getIsEmailVerified())) {
+      LOG.info("User [{}] already registered.", emailToken);
+      return;
+    }
+
+    // verify Token Expiry
+    if (emailVerificationToken.getExpiryDate().compareTo(Instant.now().toEpochMilli()) < 0) {
+      throw new RuntimeException(
+          String.format(
+              "Email Verification Token %s is expired. Please issue a new request for email verification",
+              emailVerificationToken.getToken()));
+    }
+
+    // Update the user
+    registeredUser.setIsEmailVerified(true);
+    userRepository.createOrUpdate(uriInfo, registeredUser);
+
+    // deleting the entry for the token from the Database
+    tokenRepository.deleteTokenByUserAndType(registeredUser.getId().toString(), EMAIL_VERIFICATION.toString());
+  }
+
+  @Override
+  public void resendRegistrationToken(UriInfo uriInfo, String userName) throws IOException {
+    User registeredUser = userRepository.getByName(uriInfo, userName, userRepository.getFields("isEmailVerified"));
+    if (Boolean.TRUE.equals(registeredUser.getIsEmailVerified())) {
+      // no need to do anything
+      throw new CustomExceptionMessage(FORBIDDEN, "Email Already Verified For User.");
+    }
+    tokenRepository.deleteTokenByUserAndType(registeredUser.getId().toString(), EMAIL_VERIFICATION.toString());
+    sendEmailVerification(uriInfo, registeredUser);
+  }
+
+  @Override
+  public void sendEmailVerification(UriInfo uriInfo, User user) throws IOException {
+    if (isEmailServiceEnabled) {
```

Comment:
can we move this check to sendEmailVerification. If the service is available it will send it otherwise it won't.
Since we are going to call sendEmail in general several places such as task notification or data insights etc.. we need to repeat this many times instead add it to a central method where the email is actually being sent